src/Components/Components/src/Routing/RouteConstraint.cs

```diff
@@ -30,7 +35,7 @@ public static RouteConstraint Parse(string template, string segment, string cons
                 var newInstance = CreateRouteConstraint(constraint);
                 if (newInstance != null)
                 {
-                    _cachedConstraints[constraint] = newInstance;
+                    _cachedConstraints.TryAdd(constraint, newInstance);
```

Comment:
Comment? Explain that TryAdd will only return false when there are multiple simultaneous callers and first add winners is ok

Reply:
nit: If TryAdd returns false, you could fetch the set value in the cached constraints. That way there is only once instance of the constraint shared by all routes.

Reply:
Cool idea, updated